### PR TITLE
SONARJAVA-866 RSPEC-2167 CompareTo return value check

### PR DESCRIPTION
**Pull request by m-g-sonar**
"compareTo" return value check

- It is the sign, rather than the magnitude of the value returned from `compareTo` that matters. Returning `Integer.MIN_VALUE` does not convey a higher degree of inequality, and doing so can cause errors because the return value of `compareTo` is sometimes inversed, with the expectation that negative values become positive. However, inversing `Integer.MIN_VALUE` yields `Integer.MIN_VALUE` rather than `Integer.MAX_VALUE`.
- The rule check the return value of the `compareTo` methods and add an issue if `Integer.MIN_VALUE` is returned.

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.List;
+
+@Rule(
+  key = "S2167",
+  priority = Priority.CRITICAL,
+  tags = {"bug"})
+@BelongsToProfile(title = "Sonar way", priority = Priority.CRITICAL)
+public class CompareToReturnValueCheck extends SubscriptionBaseVisitor {
+
+  private boolean insideCompareTo = false;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java]: A simple flag will not cover the corner case where an inner class is defined inside a compareTo method of an Outer class.

File: java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD)) {
+      insideCompareTo = false;
+    }
+  }
+
+  private boolean isCompareToDeclaration(MethodTree tree) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java]: We can probably use/improve the AbstractMethodDetection mechanism to find compareTo method ( :warning: Generics are not directly supported in this class right now)

**Comment by benzonico** [comment]
I think that we should probably approach this by detecting the compareTo method then subvisit the body of the method and detecting return statements (and cutting the visit of inner classes to cover the corner case of inner classes defined inside compareTo methods).

**Comment by m-g-sonar** [comment]
Ok I'll try this approach!

**Comment by m-g-sonar** [comment]
Updated code to use BaseTreeVisitor.

File: java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (isCompareToDeclaration(methodTree)) {
+      methodTree.accept(new ReturnStatementVisitor(methodTree));
+    }
+  }
+
+  private boolean isCompareToDeclaration(MethodTree tree) {
+    return isMethodName(tree, "compareTo") && hasOneNonPrimitiveParameter(tree) && returnsInt(tree);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java]: Method has only one purpose, find out name of method is "compareTo" so this parameter should be part of the method and the method should be renamed.

File: java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    public ReturnStatementVisitor(MethodTree callingMethod) {
+      this.callingMethod = ((MethodTreeImpl) callingMethod).getAstNode();
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      if (isWithinCallingMethodBody(tree) && returnsIntegerMinValue(tree.expression())) {
+        addIssue(tree, "Simply return -1");
+      }
+      super.visitReturnStatement(tree);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java]: uncessary call to super. No need to visit further.

**Comment by m-g-sonar** [comment]
Code updated.

File: java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    if (isCompareToDeclaration(methodTree)) {
+      methodTree.accept(new ReturnStatementVisitor());
+    }
+  }
+
+  private boolean isCompareToDeclaration(MethodTree tree) {
+    return isCalledCompareTo(tree) && hasOneNonPrimitiveParameter(tree) && returnsInt(tree);
+  }
+
+  private boolean isCalledCompareTo(MethodTree tree) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CompareToReturnValueCheck.java]: named would be better : the term call has a total different meaning in the context of java code.

**Comment by m-g-sonar** [comment]
Method renamed with more appropriate name.

**Comment by benzonico** [comment]
LGTM.

**Pull request closed without merging**